design/ceph-device-set.md

```diff
@@ -0,0 +1,86 @@
+  # The Ceph cluster these devices belong to
+  cluster: my-ceph-cluster
+  # Number of devices in this set
+  count: 3
+  # Capacity of each device
+  capacity: 1Ti
+  # StorageClass to provision the storage devices
+  storageClassName: aws-ebs
+  # Resource requests/limits to apply to the OSD pods
+  resources:
+    requests:
+      cpu: 2000m
+      memory: 4Gi
+    limits:
+      cpu: 2000m
+      memory: 4Gi
+  # Placement constraints to apply to the OSD pods
+  placement:
+    nodeAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+        nodeSelectorTerms:
+        - matchExpressions:
+          - key: "failure-domain.beta.kubernetes.io/zone"
+            operator: In
+            values:
+            - us-west-1a
+    podAntiAffinity:
+      preferredDuringSchedulingIgnoredDuringExecution:
+      - weight: 100
+        podAffinityTerm:
+          labelSelector:
+            matchExpressions:
+            - key: "ocs.openshift.io/cluster"
+              operator: In
+              values:
+              - my-ceph-cluster
+          topologyKey: "failure-domain.beta.kubernetes.io/zone"
+```
+
+The CephDeviceSet CRD will define characteristics of a set of OSD devices and their Pods. The following fields are available in the spec:
+
+* **cluster**: The Ceph cluster the devices will belong to. **[required]**
+* **count**: The number of devices in the set. **[required]**
+* **capacity**: The capacity of the devices in this set. **[required]**
+* **storageClassName**: The StorageClass to provision the PVCs from. If not specified, the cluster default StorageClass will be used.
+* **resources**: The CPU and RAM requests/limits to apply to the OSD Pods. Default is no resource requests.
+* **placement**: The placement criteria to apply to the OSD Pods. Default is no placement criteria.
+
+## CephCluster CRD Requirements
+The CephCluster CRD will have to be extended to include a new field: `spec.storage.useDeviceSets`. If this is `true`, the CephCluster controller will skip the usual task of creating OsdPrepare Jobs and OSD Deployments. Configuring either `useAllDevices` or specific devices on nodes will be mutually exclusive from this new configuration option. Default will be `false`, to preserve current behavior.
+
+## Workflow
+To start, the admin will need to create a CephCluster CR with `spec.storage.useDeviceSets` set to `true`. This will cause the CephCluster controller to not discover, prepare, or create OSDs. The admin will then create one or more CephDeviceSet CRs to specify the storage for the cluster.
+
+If the admin wishes to control device placement beyond simply using node hostnames it will be up to them to make sure the desired nodes are labeled properly to ensure the Kuberenetes scheduler will distribute the OSD Pods appropriately.
+
+Additionally, it is recommended that the admin setup a StorageClass with `volumeBindingMode: WaitForFirstConsumer` set, and that they use this StorageClass for providing the OSD PVCs.
```

Comment:
```suggestion
Additionally, it is recommended that the admin setups a StorageClass with `volumeBindingMode: WaitForFirstConsumer` set, and that they use this StorageClass for providing the OSD PVCs.
```

Reply:
This change doesn't make grammatical sense... unless you mean to replace it with `sets up`, but that's not quite grammatically correct in English either.

Reply:
Ok never mind then.